controller/watch/watcher_test.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watch
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"k8s.io/client-go/1.5/pkg/api/unversioned"
+	"k8s.io/client-go/1.5/pkg/api/v1"
+	extv1beta1 "k8s.io/client-go/1.5/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/1.5/pkg/watch"
+)
+
+func TestDeploymentWatcher(t *testing.T) {
+	const (
+		evtChTimeout = 100 * time.Millisecond
+	)
+	fakeWatcher := watch.NewFake()
+	deplIface := &fakeDeploymentInterface{
+		listRet: &extv1beta1.DeploymentList{
+			TypeMeta: unversioned.TypeMeta{},
+			ListMeta: unversioned.ListMeta{},
+			Items: []extv1beta1.Deployment{
+				{ObjectMeta: v1.ObjectMeta{Name: "listdepl1"}},
+				{ObjectMeta: v1.ObjectMeta{Name: "listdepl2"}},
+			},
+		},
+		watchRet: fakeWatcher,
+	}
+	evtCh := make(chan watch.Event)
+	wcb := func(evt watch.Event) error {
+		evtCh <- evt
+		return nil
+	}
+	go deploymentWatcher(deplIface, wcb)
+
+	numRecv := 0
+	numItems := len(deplIface.listRet.Items)
+	done := false
+	// ensure exactly the list of deployments is processed first
+	for {
+		select {
+		case evt := <-evtCh:
+			if numRecv >= numItems {
```

Comment:
I don't believe so, because `numRecv` is zero-based (effectively the index into `deplIface.listRet.Items`)